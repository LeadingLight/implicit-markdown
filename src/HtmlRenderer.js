
const renderRules = {
  paragraph_open: (tokens, idx) => (tokens[idx].tight ? '' : '<p>'),
  paragraph_close: (tokens, idx) => {
    const addBreak = !(tokens[idx].tight && idx && tokens[idx - 1].type === 'inline' && !tokens[idx - 1].content);

    return (tokens[idx].tight ? '' : '</p>') + (addBreak ? getBreak(tokens, idx) : '');
  },
  text: (tokens, idx) => tokens[idx].content,
  heading_open: (tokens, idx /* , options, env */) => `<h${tokens[idx].hLevel}>`,
  heading_close: (tokens, idx /* , options, env */) => `</h${tokens[idx].hLevel}>\n`
};


function nextToken(tokens, idx) {
  if (++idx >= tokens.length - 2) {
    return idx;
  }
  if (tokens[idx].type === 'paragraph_open' && tokens[idx].tight
    && (tokens[idx + 1].type === 'inline' && tokens[idx + 1].content.length === 0)
    && (tokens[idx + 2].type === 'paragraph_close' && tokens[idx + 2].tight)) {
    return nextToken(tokens, idx + 2);
  }

  return idx;
}

function getBreak(tokens, idx) {
  idx = nextToken(tokens, idx);
  if (idx < tokens.length && tokens[idx].type === 'list_item_close') {
    return '';
  }

  return '\n';
}

export default class Renderer {
  constructor() {
    this.rules = renderRules;
  }

  renderInline(tokens, options, env) {
    const _rules = this.rules;
    let len = tokens.length, i = 0;
    let result = '';

    while (len--) {
      result = result + _rules[tokens[i].type](tokens, i++, options, env, this);
    }

    return result;
  }

  render(tokens, options, env) {
    const _rules = this.rules;
    let len = tokens.length, i = -1;
    let result = '';

    while (++i < len) {
      if (tokens[i].type === 'inline') {
        result = result + this.renderInline(tokens[i].children, options, env);
      } else {
        result = result + _rules[tokens[i].type](tokens, i, options, env, this);
      }
    }

    return result;
  }
}
