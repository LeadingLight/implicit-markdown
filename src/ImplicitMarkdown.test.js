import ImplicitMarkdown from './index';
import HtmlRenderer from './HtmlRenderer';

const JSON_INDENT = 2;

describe('ImplicitMarkdown', () => {
  it('should parse simple string to a json object renderable by implicit-ui', () => {
    const md = new ImplicitMarkdown(new HtmlRenderer());
    const mdTextWithLink = 'this is a text with a [link](#this-is-an-anchor) in it';

    expect(md.parse(mdTextWithLink)).toMatchSnapshot();
  });

  it('test paragraph render', () => {
    const md = new ImplicitMarkdown(new HtmlRenderer());
    const mdTextWithLink = 'this is a text with a [link](#this-is-an-anchor) in it';

    expect(md.render(mdTextWithLink)).toMatchSnapshot();
  });

  it('testing headers parse', () => {
    const md = new ImplicitMarkdown(new HtmlRenderer());
    const mdTextWithLink = '# this is a title';

    expect(md.parse(mdTextWithLink)).toMatchSnapshot();
  });

  it('testing headers render', () => {
    const md = new ImplicitMarkdown(new HtmlRenderer());
    const mdTextWithLink = '# this is a title';

    expect(md.render(mdTextWithLink)).toMatchSnapshot();
  });

  it('empty', () => {
    const md = new ImplicitMarkdown(new HtmlRenderer());

    expect(md.render('')).toMatchSnapshot();
  });
});

