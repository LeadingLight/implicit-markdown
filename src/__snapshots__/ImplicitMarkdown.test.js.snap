// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImplicitMarkdown should parse simple string to a json object renderable by implicit-ui 1`] = `
Array [
  Object {
    "level": 0,
    "lines": Array [
      0,
      1,
    ],
    "tight": false,
    "type": "paragraph_open",
  },
  Object {
    "children": Array [
      Object {
        "content": "this is a text with a [link](#this-is-an-anchor) in it",
        "level": 0,
        "type": "text",
      },
    ],
    "content": "this is a text with a [link](#this-is-an-anchor) in it",
    "level": 1,
    "lines": Array [
      0,
      1,
    ],
    "type": "inline",
  },
  Object {
    "level": 0,
    "tight": false,
    "type": "paragraph_close",
  },
]
`;

exports[`ImplicitMarkdown test paragraph render 1`] = `
"<p>this is a text with a [link](#this-is-an-anchor) in it</p>
"
`;

exports[`ImplicitMarkdown testing headers parse 1`] = `
Array [
  Object {
    "hLevel": 1,
    "level": 0,
    "lines": Array [
      0,
      1,
    ],
    "type": "heading_open",
  },
  Object {
    "children": Array [
      Object {
        "content": "this is a title",
        "level": 0,
        "type": "text",
      },
    ],
    "content": "this is a title",
    "level": 1,
    "lines": Array [
      0,
      1,
    ],
    "type": "inline",
  },
  Object {
    "hLevel": 1,
    "level": 0,
    "type": "heading_close",
  },
]
`;

exports[`ImplicitMarkdown testing headers render 1`] = `
"<h1>this is a title</h1>
"
`;
